stdlib/public/core/ArraySlice.swift

```diff
@@ -0,0 +1,70 @@
+
+public typealias ArraySlice<T> = Slice<[T]>
+
+extension Slice: _ContiguouslyStored where Base: _ContiguouslyStored
+/* FIXME: exterminate */, Base.IndexDistance == Int 
+{
+  public func withUnsafeBufferPointer<R>(
+    _ body: (UnsafeBufferPointer<Element>) throws -> R
+  ) rethrows -> R {
+    let i = _base.distance(from: _base.startIndex, to: _startIndex)
+    let j = _base.distance(from: _base.startIndex, to: _endIndex)
+    return try _base.withUnsafeBufferPointer { buf in
+      let sliced = UnsafeBufferPointer(rebasing: buf[i..<j])
+      return try body(sliced)
+    }
+  }
+
+  public mutating func withUnsafeMutableBufferPointer<R>(
+    _ body: (inout UnsafeMutableBufferPointer<Element>) throws -> R
+  ) rethrows -> R {
+    let i = _base.distance(from: _base.startIndex, to: _startIndex)
+    let j = _base.distance(from: _base.startIndex, to: _endIndex)
+    return try _base.withUnsafeMutableBufferPointer { buf in
+      var sliced = UnsafeMutableBufferPointer(rebasing: buf[i..<j])
+      return try body(&sliced)
+    }
+  }
+
+  @_inlineable
+  public mutating func withUnsafeMutableBytes<R>(
+    _ body: (UnsafeMutableRawBufferPointer) throws -> R
+  ) rethrows -> R {
+    return try self.withUnsafeMutableBufferPointer {
+      return try body(UnsafeMutableRawBufferPointer($0))
+    }
+  }
+  @_inlineable
+  public func withUnsafeBytes<R>(
+    _ body: (UnsafeRawBufferPointer) throws -> R
+  ) rethrows -> R {
+    return try self.withUnsafeBufferPointer {
+      try body(UnsafeRawBufferPointer($0))
+    }
+  }
+}
```

Comment:
Could we also do a faster `_copyToContiguousArray()` and `_copyContents(initializing:)` if we know the base collection is also contiguous?

Reply:
Yes. My hope is that we can formalize those to be official API and drop their underscores too, as part of a proposal for the new protocol.